src/ripple/app/hook/impl/applyHook.cpp

```diff
@@ -0,0 +1,4976 @@
+            {
+                if (a == 0 || b == 0 || c == 0 || d == 0)
+                   return INVALID_ARGUMENT;
+
+
+                uint32_t aread_ptr = a, aread_len = b, kread_ptr = c, kread_len = d, nread_ptr = e, nread_len = f;
+
+                if (NOT_IN_BOUNDS(aread_ptr, aread_len, memory_length) ||
+                    NOT_IN_BOUNDS(kread_ptr, kread_len, memory_length) ||
+                    NOT_IN_BOUNDS(nread_ptr, nread_len, memory_length))
+                   return OUT_OF_BOUNDS;
+
+                if (aread_len != 20 || kread_len != 32 || nread_len != 32)
+                    return INVALID_ARGUMENT;
+
+                ripple::Keylet kl =
+                    ripple::keylet::hookState(
+                            ripple::base_uint<160, ripple::detail::AccountIDTag>::fromVoid(memory + aread_ptr),
+                            ripple::base_uint<256>::fromVoid(memory + kread_ptr),
+                            ripple::base_uint<256>::fromVoid(memory + nread_ptr));
+
+                return serialize_keylet(kl, memory, write_ptr, write_len);
+            }
+
+
+            // skip is overloaded, has a single, optional 4 byte argument
+            case keylet_code::SKIP:
+            {
+                if (c != 0 || d != 0 || e != 0 || f != 0)
+                   return INVALID_ARGUMENT;
+
+                ripple::Keylet kl =
+                    (b == 0 ? ripple::keylet::skip() :
+                    ripple::keylet::skip(a));
+
+                return serialize_keylet(kl, memory, write_ptr, write_len);
+            }
+
+            // no arguments
+            case keylet_code::AMENDMENTS:
+            case keylet_code::FEES:
+            case keylet_code::NEGATIVE_UNL:
+            case keylet_code::EMITTED_DIR:
+            {
+                if (a != 0 || b != 0 || c != 0 || d != 0 || e != 0 || f != 0)
+                   return INVALID_ARGUMENT;
+
+                ripple::Keylet kl =
+                    keylet_type == keylet_code::AMENDMENTS   ? ripple::keylet::amendments()       :
+                    keylet_type == keylet_code::FEES         ? ripple::keylet::fees()             :
+                    keylet_type == keylet_code::NEGATIVE_UNL ? ripple::keylet::negativeUNL()      :
+                    ripple::keylet::emittedDir();
+
+                return serialize_keylet(kl, memory, write_ptr, write_len);
+            }
+
+            case keylet_code::LINE:
+            {
+                if (a == 0 || b == 0 || c == 0 || d == 0 || e == 0 || f == 0)
+                   return INVALID_ARGUMENT;
+
+                uint32_t hi_ptr = a, hi_len = b, lo_ptr = c, lo_len = d, cu_ptr = e, cu_len = f;
+
+                if (NOT_IN_BOUNDS(hi_ptr, hi_len, memory_length) ||
+                    NOT_IN_BOUNDS(lo_ptr, lo_len, memory_length) ||
+                    NOT_IN_BOUNDS(cu_ptr, cu_len, memory_length))
+                   return OUT_OF_BOUNDS;
+
+                if (hi_len != 20 || lo_len != 20 || cu_len != 20)
```

Comment:
As you know the bounds check is looking for attempted illegal memory access, in particular access outside the Hook's designated memory. I've tried to be fairly consistent in placing these as near the top of each API implementation (or in this case sub implementation) as possible to make it visually easy to see if the bounds are being checked correctly.

Passing the static sizes has more to do with having a consistent hooks API ([see doco](https://xrpl-hooks.readme.io/reference/hook-api-conventions#parameters)) wherein pointers are never passed without a size to accompany them. That is: you always pass information about a buffer in your hook, never just a pointer. It makes the developer stop and think about their memory, and prevents silly errors which would clobber their wasm stack frame, produce difficult to debug errors, and possibly even open vulnerabilities in their hook. It also makes the API forward compatible, if these sizes ever change or we need to overload a function somehow we can use them. (Case in point: the seqID amendment I will be putting forward shortly allows some keylets to be built using a uint256 instead of an acc seq.) Finally it makes developing hooks easier using macros like SBUF. These are the pros, the con is passing essentially static values in order to use this API. In my mind the pros win, but willing to discuss further if you like